polytracker/include/taintdag/taint_source.hpp

```diff
@@ -0,0 +1,84 @@
+#ifndef TDAG_TAINT_SOURCE_HPP
+#define TDAG_TAINT_SOURCE_HPP
+
+#include <limits>
+#include <optional>
+
+#include "taintdag/section.hpp"
+#include "taintdag/string_table.hpp"
+#include "taintdag/taint.hpp"
+
+namespace taintdag {
+
+struct SourceEntry {
+  // Relative to StringTable memory start
+  StringTable::offset_t string_offset;
+  int fd;
+
+  std::string_view name(StringTable const &st) const {
+    return st.from_offset(string_offset);
+  }
+};
+
+// NOTE(hbrodin): Ideally this would be separated into sources and sinks.
+// The issue is that we don't really know until there is a write to a fd
+// that it was a sink (even if opened writeable).
+// TODO(hbrodin): Consider creating a separate Sinks structure that is used
+// when a file is opened for writing (signalling that it could be a sink).
+// That would also apply to when creating another bidirectional fd such as
+// sockets. For argv or files opened readonly only a source would be created.
+struct Sources : public FixedSizeAlloc<SourceEntry> {
+  // Index of source, maximum number of sources is limited by this
+  using index_t = source_index_t;
+
+  static constexpr uint8_t tag{1};
+  static constexpr size_t allocation_size{std::numeric_limits<index_t>::max() *
+                                          sizeof(SourceEntry)};
+
+  template <typename OF>
+  Sources(SectionArg<OF> of)
+      : FixedSizeAlloc{of.range},
+        st_{of.output_file.template section<StringTable>()} {
+    // TODO(hbrodin): Drop the assert, replace with error_exit.
```

Comment:
Not sure I follow. I would just like to remove the assert and replace it with the 'error_exit'-call in case of error. For testing, one could replace the `error_exit`-function, given that we have exceptions enable and that way ensure it works as expected.
The if-we-have-exception question is still open as we do compile with exceptions disabled it seems. That is one if the issues we have getting the test cases to run as expected.

Reply:
gotcha! I was just wondering if you'd tried that and it hadn't worked for some reason hence the note - compiling with exceptions disabled would explain it!